core/domain/classifier_services.py

```diff
@@ -424,41 +375,240 @@ def delete_classifier_training_job(job_id):
+    combination.
+
+    Args:
+        exp_id: str. ID of the exploration.
+        exp_version: int. Version of the exploration.
+        state_name: str. Name of the state for which training job mapping model
+            is to be retrieved.
+
+    Returns:
+        StateTrainingJobsMapping. A domain object containing exploration
+        mapping model information.
+    """
+    state_training_jobs_mapping_model = (
+        classifier_models.StateTrainingJobsMappingModel.get_model(
+            exp_id, exp_version, state_name))
+    if state_training_jobs_mapping_model:
+        return classifier_domain.StateTrainingJobsMapping(
+            state_training_jobs_mapping_model.exp_id,
+            state_training_jobs_mapping_model.exp_version,
+            state_training_jobs_mapping_model.state_name,
+            state_training_jobs_mapping_model.algorithm_ids_to_job_ids)
+
+
+def migrate_exploration_training_jobs(state_training_jobs_mapping):
+    """Migrate exploration training jobs to latest version of algorithm_id
+    and algorithm_version.
+
+    This function lazily migrates an older classifier training job and
+    trains new classifiers. Specifically, if training job exploration mapping of
+    an <exploration, version, state> triplet is missing job_id for some
+    algorithm_id, or if the job_id exists but it has been trained on a now
+    obsolete algorithm, we re-submit the jobs.
+
+    The function goes through existing training job exploration mapping and
+    identifies three different types of algorithm IDs.
+        1. algorithm_ids_to_upgrade: Those which exist but needs to be
+            upgraded a new algorithm (because existing one has been deprecated)
+            by re-submitting the training job.
+        2. algorithm_ids_to_add: Those which doesn't exist and needs to be added
+            by submitting a new training job.
+        3. algorithm_ids_to_remove: Those which needs to be removed since these
+            algorithms are no longer supported.
+
+    Once all three types of algorithm IDs are filtered, the function performs
+    specific tasks tailored to each of them. Its called lazy migration because
+    it happens only when there is a query to retrieve trained model for given
+    <exploration, version, state> and algorithm_id.
+
+    Args:
+        state_training_jobs_mapping: StateTrainingJobsMapping. Domain
+            object containing exploration to training job id mapping. This
+            mapping is used to figure out jobs that needs to be re-submitted /
+            added or removed.
+    """
+    exp_id = state_training_jobs_mapping.exp_id
+    exp_version = state_training_jobs_mapping.exp_version
+    state_name = state_training_jobs_mapping.state_name
+
+    exploration = exp_fetchers.get_exploration_by_id(
+        exp_id, version=exp_version)
+    interaction_id = exploration.states[state_name].interaction.id
+
+    algorithm_id = feconf.INTERACTION_CLASSIFIER_MAPPING[
```

Comment:
Sorry -- I think I'm missing something again. I thought an interaction can have multiple algorithm IDs? At least, that's what it says here: https://github.com/oppia/oppia/pull/10060/files#diff-d92aa6c527ae9e47b02fcd58593b875aR371

Reply:
Yes, it can, but right now, we are going ahead with single ID mechanism, hence only single algorithm_id. In future, when we do have cases for multiple algrotihm IDs this will be a list rather than a single ID. See relevant discussion [here](https://docs.google.com/document/d/1fIlDem4NC6eqEHwjkn2FNdHuo6eEsg3LMLd2QAfxCPo/edit#heading=h.7aioqynmhsel). As per our earlier discussion, the datastore models will be changed as of now so that in future when we implement the necessary application logic, we can swiftly switch over without having to migrate the datastore models. The necessary application logic is not possible because, we don't have multiple algorithms at this point.  Hence, only `StateClassifierJob` model and its domain object are changed so that in future when we do have muliple algorithm_ids, we can easily adapt to them.

Reply:
That sounds fine, but you should probably add clear comments in the relevant places to explain the difference, and maybe reframe the constant in INTERACTION_CLASSIFIER_MAPPING as DEFAULT_INTERACTION_CLASSIFIER_MAPPING or something like that. Also, please add a TODO pointing to a GitHub issue for any deferred work, and make sure that the GitHub issue fully describes what needs to be done clearly (don't just write a single-line issue).

Thanks!

Reply:
Actually, there is no issue to create, this is more like a feature in future. However I can add comment, would that be sufficient?

Reply:
Well, if we plan to do it in the future, should still create an issue, right? An issue might mean "next planned steps", it doesn't necessarily mean there's a bug. You could file it as a feature request.

(And please do definitely add comments too.)

Reply:
Done. Created issue #10217